enterprise/tools/check_metrics/check_metrics.go

```diff
@@ -0,0 +1,134 @@
+func main() {
+	flag.Parse()
+	err := yaml.PopulateFlagsFromFile(*configPath)
+	if err != nil {
+		log.Fatalf("Could not read metrics config from %s: %s", *configPath, err)
+	}
+	err = log.Configure()
+	if err != nil {
+		log.Fatalf("Could not configure logger: %s", err)
+	}
+	prometheusClient, err := api.NewClient(api.Config{
+		Address: config.PrometheusAddress,
+	})
+	if err != nil {
+		log.Fatalf("Error creating prometheus client: %s\nDid you run the prometheus port-forward command?", err)
+	}
+	promAPI := v1.NewAPI(prometheusClient)
+
+	eg, gctx := errgroup.WithContext(context.Background())
+	for _, metric := range config.PrometheusMetrics {
+		metric := metric
+		eg.Go(func() error {
+			monitoringTimer := time.After(time.Duration(config.MonitoringTimeframeSeconds) * time.Second)
+			monitoringTicker := time.NewTicker(time.Duration(metric.PollingIntervalSeconds) * time.Second)
+			defer monitoringTicker.Stop()
+
+			unhealthyCount := 0
+			for {
+				select {
+				case <-gctx.Done():
+					return nil
+				case <-monitoringTimer:
+					return nil
+				case <-monitoringTicker.C:
+					healthy, err := metricHealthy(promAPI, metric.Name, metric.CanaryQuery, metric.BaselineQuery, metric.HealthThreshold)
+					if err != nil {
+						log.Warningf("Error querying metric %s: %s", metric.Name, err)
+					}
+
+					if healthy {
+						unhealthyCount = 0
+					} else {
+						unhealthyCount++
+					}
+
+					if unhealthyCount >= metric.MaxUnhealthyCount {
+						return errors.New(fmt.Sprintf("%s metrics unhealthy.", metric.Name))
+					}
+				}
+			}
+		})
+	}
+
+	err = eg.Wait()
+	if err != nil {
+		log.Fatalf("Exiting metrics script: %s", err)
+	}
+
+	// If all metrics look healthy for the duration of the monitoring timeframe, continue with rollout
+	os.Exit(metricsHealthyExitCode)
+}
+
+// Returns whether the given metric is healthy compared to the baseline
+// If the metric's success rate is lower than the baseline's by the input threshold, this will return false
+// Otherwise will return true
+func metricHealthy(v1api v1.API, metricName string, query string, baselineQuery string, healthThreshold float64) (bool, error) {
+	ctx := context.Background()
+	log.Debugf("Querying metric %s", metricName)
+
+	result, _, err := v1api.Query(ctx, query, time.Now())
+	if err != nil {
+		return false, err
+	}
+	resultVector := result.(model.Vector)
+	metricValue := model.SampleValue(0)
+	if len(resultVector) > 0 {
+		metricValue = resultVector[0].Value
+	}
+
+	baselineResult, _, err := v1api.Query(ctx, baselineQuery, time.Now())
+	if err != nil {
+		return false, err
+	}
+	baselineResultVector := baselineResult.(model.Vector)
+	baselineMetricValue := model.SampleValue(0)
+	if len(baselineResultVector) > 0 {
+		baselineMetricValue = baselineResultVector[0].Value
+	}
+
+	// If the metric differs from the baseline by the set threshold, the metric is considered unhealthy
```

Comment:
since we're doing Abs() here could we get into a weird state where the canary is actually more healthy than the other apps and we're reporting it as unhealthy?

Reply:
Yeah... I thought about that too. I guess that could happen if we fix an error, so the canary has fewer GRPC errors than the other apps for example

Tried to clarify this, but let me know if I just made it more confusing

Reply:
Refactored the way to specify health thresholds. Hopefully it's a lot more clearer now